libz-rs-sys/src/gz.rs

```diff
@@ -301,3 +468,255 @@ pub unsafe extern "C-unwind" fn gzclose(file: gzFile) -> c_int {
         Z_ERRNO
     }
 }
+
+/// Retrieve the zlib error code and a human-readable string description of
+/// the most recent error on a gzip file stream.
+///
+/// # Arguments
+///
+/// * `file` - A gzip file handle, or null
+/// * `errnum` - A pointer to a C integer in which the zlib error code should be
+///   written, or null if the caller does not need the numeric error code.
+///
+/// # Returns
+///
+/// * A pointer to a null-terminated C string describing the error, if `file` is non-null
+///   and has an error
+/// * A pointer to an empty (zero-length), null-terminated C string, if `file` is non-null
+///   but has no error
+/// * Null otherwise
+///
+/// # Safety
+///
+/// `file` must be one of the following:
+/// - A file handle obtained from [`gzopen`] or [`gzdopen`].
+/// - A null pointer.
+///
+/// If this function returns a non-null string, the caller must not modifiy or
+/// deallocate the string.
+///
+/// If `errnum` is non-null, it must point to an address at which a [`c_int`] may be written.
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzerror))]
+pub unsafe extern "C-unwind" fn gzerror(file: gzFile, errnum: *mut c_int) -> *const c_char {
+    // Get internal structure and check integrity
+    if file.is_null() {
+        return ptr::null();
+    }
+    let Some(state) = (unsafe { file.cast::<GzState>().as_ref() }) else {
+        return ptr::null();
+    };
+    if state.mode != GzMode::GZ_READ && state.mode != GzMode::GZ_WRITE {
+        return ptr::null();
+    }
+
+    // Return error information
+    if !errnum.is_null() {
+        // Safety:
+        // * `errnum` is non-null
+        // * The caller is responsible for ensuring that `errnum` points to writable
+        //   memory with proper alignment.
+        unsafe { *errnum = state.err };
+    }
+    if state.err == Z_MEM_ERROR {
+        b"out of memory\0".as_ptr().cast::<c_char>()
+    } else if state.msg.is_null() {
+        b"\0".as_ptr().cast::<c_char>()
+    } else {
+        state.msg
+    }
+}
+
+// Create a deep copy of a C string using `ALLOCATOR`
+//
+// # Safety
+//
+// The caller must ensure that s is either null or a pointer to a null-terminated C string.
+unsafe fn gz_strdup(src: *const c_char) -> *mut c_char {
+    if src.is_null() {
+        return ptr::null_mut();
+    }
+
+    // SAFETY: the caller must ensure this is a valid C string
+    let src = unsafe { CStr::from_ptr(src) };
+
+    let len = src.to_bytes_with_nul().len();
+    let Some(dst) = ALLOCATOR.allocate_slice_raw::<c_char>(len) else {
+        return ptr::null_mut();
+    };
+
+    // SAFETY: src and dst don't overlap, because dst was just allocated. src is valid for a read
+    // of len bytes, and dst is valid for a write of len bytes.
+    unsafe { core::ptr::copy_nonoverlapping(src.as_ptr(), dst.as_ptr(), len) };
```

Comment:
I'm using the `core::ptr` function here (and not some libc function) so that this runs with Miri